src/librustc_mir/borrow_check.rs

```diff
@@ -599,12 +607,24 @@ impl<'c, 'b, 'a: 'b+'c, 'gcx, 'tcx: 'a> MirBorrowckCtxt<'c, 'b, 'a, 'gcx, 'tcx>
         let ty = lvalue.ty(self.mir, self.tcx).to_ty(self.tcx);
         let moves_by_default =
             self.fake_infer_ctxt.type_moves_by_default(self.param_env, ty, DUMMY_SP);
-        if moves_by_default {
-            // move of lvalue: check if this is move of already borrowed path
-            self.access_lvalue(context, lvalue_span, (Deep, Write(WriteKind::Move)), flow_state);
-        } else {
-            // copy of lvalue: check if this is "copy of frozen path" (FIXME: see check_loans.rs)
-            self.access_lvalue(context, lvalue_span, (Deep, Read(ReadKind::Copy)), flow_state);
+
+        // Check if error has already been reported to stop duplicate reporting.
+        let has_storage_drop_or_dead_error_reported = match *lvalue {
+            Lvalue::Local(local) => self.storage_drop_or_dead_error_reported.insert(local),
+            _ => false,
+        };
+
+        if !has_storage_drop_or_dead_error_reported {
```

Comment:
So @arielb1 pointed out that there is a risk here. The calls to `access_lvalue` are not guaranteed to report an error -- and, in fact, DROP will report errors in a broader set of conditions than `StorageDead` (I think strictly broader). So it would be better to populate the map (and hence skip the latter check) only if an error was actually reported.

What @arielb1 suggested was issuing precisely the same error message in both cases, and letting rustc deduplicate (which it will do). That's an option, though I don't think it's the best path, since I think that the error messages should perhaps be different between the two cases.

Alternatively, we could make `access_lvalue` return a boolean or something, indicating whether an error was reported, and then do:

```
let has_storage_drop_or_dead_error_reported = self.storage_drop_or_dead_error_reported.contains_key(...);

let error_reported;
if has_storage_drop_or_dead_error_reported { error_reported = false; } else {
    if moves_by_default { error_reported = self.access_lvalue(..); } else { ... }
}
if error_reported { self.storage_drop_or_dead_error_reported.insert(local); }
```

It would also be good to comment about this. =)

It's hard to come up with a test case that would necesarily show off this problem though -- specifically, `Drop` will always come before `StorageDead`, and hence I imagine we'll visit it first, so the bug won't show itself. But I wouldn't want to rely on that.

Reply:
I've modified what was there based on this, should be fine now.